### PR TITLE
JUnit 5 support

### PR DESCRIPTION
**Pull request by pativa**
This PR implements support for JUnit 5 in `ymer-test`.

* JUnit 4 tests are moved to `com.avanza.ymer.junit4` package.
* The previous test classes are still available as @Deprecated classes, intended to be replaced by either JUnit 4 or JUnit 5 implementation.
* New, functionally identical JUnit 5 tests added in `com.avanza.ymer.junit5` package.
* `junit` (JUnit 4) and `junit-jupiter-api` + `junit-jupiter-params` (JUnit 5) dependencies are optional. The application should include the one to be used.

This PR takes a lot of inspiration in the implementation from https://github.com/AvanzaBank/gs-test/pull/28

### CODE REVIEW

**Comment by ath0s** [comment]
Looks good!
I'm wonder if it would be worth the trouble to introduce separate modules for JUnit4 and 5 instead. An example of this can be found here https://github.com/AvanzaBank/astrix/pull/108
Some advantages would be that we don't have to deprecate anything, we would be redirecting `ymer-test` to `ymer-test-junit4` (or however it would be named) instead. The JUnit4 and JUnit5 modules could have dependencies to the appropriate version without making them optional.
Disadvantages include introduces multiple pretty small modules.

**Comment by pativa** [comment]
I think you might be right. Not deprecating anything is a big advantage, so everyone depending on this don't have to get warnings. I'll look into splitting this to separate modules.

File: examples/example-pu/pom.xml [hunk 1]
```diff
@@ -19,13 +19,18 @@
...
 		<dependency>
-			<groupId>junit</groupId>
-			<artifactId>junit</artifactId>
+			<groupId>org.junit.jupiter</groupId>
+			<artifactId>junit-jupiter-engine</artifactId>
+			<scope>test</scope>
+		</dependency>
+		<dependency>
+			<groupId>org.junit.jupiter</groupId>
+			<artifactId>junit-jupiter-params</artifactId>
```
Comment (ath0s) [inline examples/example-pu/pom.xml:33]: I suggest just adding a test dependency to `junit-jupiter`, it includes `-api`, `-engine` and `-params`

File: ymer-test-junit5/pom.xml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+			<artifactId>hamcrest</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>org.hamcrest</groupId>
+			<artifactId>hamcrest-library</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>org.hamcrest</groupId>
+			<artifactId>hamcrest-core</artifactId>
+		</dependency>
```
Comment (ath0s) [inline ymer-test-junit5/pom.xml:35]: `<artifactId>hamcrest</artifactId>` is enough, `-library` and `-core` are basically just aliases for that one (from version 2.1).
See http://hamcrest.org/JavaHamcrest/distributables#previous-versions-of-hamcrest

File: examples/example-pu/pom.xml (hunk 1 above)
Comment (pativa) [reply examples/example-pu/pom.xml:33]: Well spotted! I didn't check the changes in the example app too closely. I'll fix that!

File: ymer-test-junit4/pom.xml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+			<artifactId>hamcrest-library</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>org.hamcrest</groupId>
+			<artifactId>hamcrest-core</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>junit</groupId>
+			<artifactId>junit</artifactId>
+		</dependency>
```
Comment (ath0s) [inline ymer-test-junit4/pom.xml:31]: You don't need all of these.
I would include only `<artifactId>hamcrest</artifactId>`, and exclude `<artifactId>hamcrest-core</artifactId>` from the junit dependency:
```suggestion
		<dependency>
			<groupId>org.hamcrest</groupId>
			<artifactId>hamcrest</artifactId>
		</dependency>		
		<dependency>
			<groupId>junit</groupId>
			<artifactId>junit</artifactId>
			<exclusions>
				<exclusion>
					<groupId>org.hamcrest</groupId>
					<artifactId>hamcrest-core</artifactId>
				</exclusion>
			</exclusions>
		</dependency>
```

File: ymer-test-junit5/pom.xml (hunk 2 above)
Comment (pativa) [reply ymer-test-junit5/pom.xml:35]: Nice! I just copied the dependencies, but this is a good opportunity for some cleaning as well!

File: pom.xml [hunk 4]
```diff
@@ -229,12 +233,24 @@
...
 				<groupId>com.avanza.gs</groupId>
 				<artifactId>gs-test</artifactId>
 				<version>${gs-test.version}</version>
 				<scope>test</scope>
 			</dependency>
+			<dependency>
+				<groupId>com.tngtech.archunit</groupId>
+				<artifactId>archunit-junit5</artifactId>
+				<version>${archunit.version}</version>
+			</dependency>
```
Comment (ath0s) [inline pom.xml:253]: I don't see ArchUnit being used, to I think the dependency could be removed.

**Review by ath0s — APPROVED** [review]

File: pom.xml (hunk 4 above)
Comment (pativa) [reply pom.xml]: Fixed!

**Review by Bjornelmers — APPROVED** [review]
Looks good!

**Pull request merged**